code/modules/maze_generation/maze_generator.dm

```diff
@@ -0,0 +1,174 @@
+	/// Total turfs
+	var/total_turfs = 0
+	/// Generation start time
+	var/start_time = 0
+	/// Maze width
+	var/mwidth = 0
+	/// Maze height
+	var/mheight = 0
+
+/obj/effect/mazegen/generator/Initialize(mapload)
+	. = ..()
+	SSlate_mapping.maze_generators += src
+
+/obj/effect/mazegen/generator/Destroy()
+	SSlate_mapping.maze_generators -= src
+	return ..()
+
+// "Push" a turf to the working "stack"
+/obj/effect/mazegen/generator/proc/push_turf(turf/T)
+	working_stack.Add(T) // Add it to the end of the list
+
+// "Pop" a turf off the working "stack"
+/obj/effect/mazegen/generator/proc/pop_turf()
+	var/turf/T = working_stack[length(working_stack)] // Get the last item in the list
+	working_stack.Remove(T) // Take it off the top
+	return T // Send it back
+
+// This is the main proc that does the work. It can get very performance intensive if youre not careful
+// I was testing this and I managed to make the MC erase itself. Dont ask.
+/obj/effect/mazegen/generator/proc/generate()
+	// Setup some initial vars
+	var/start_time = start_watch()
+	log_debug("Starting to generate maze at [x],[y],[z] (WxH: [mwidth]x[mheight])")
+
+	// Setup our turf list
+	var/width = clamp(mwidth, 0, (world.maxx - x)) // Make sure they dont loop off the world
+	var/height = clamp(mheight, 0, (world.maxy - y)) // Make sure they dont loop off the world
+
+	// Generate a turf stack
+	for(var/target_x in 0 to (width - 1)) // -1 so it spawns on the right tile
+		for(var/target_y in 0 to (height - 1))
+			var/turf/T = locate((x + target_x), (y + target_y), z)
+			// Mark as unsearched
+			T.color = MAZEGEN_TURF_UNSEARCHED
+			// Throw it in the list
+			turf_list |= T
+
+			// Windows time
+			var/obj/structure/window/reinforced/mazeglass/WN = new(T)
+			WN.dir = NORTH
+			var/obj/structure/window/reinforced/mazeglass/WS = new(T)
+			WS.dir = SOUTH
+			var/obj/structure/window/reinforced/mazeglass/WE = new(T)
+			WE.dir = EAST
+			var/obj/structure/window/reinforced/mazeglass/WW = new(T)
+			WW.dir = WEST
+			CHECK_TICK
+		CHECK_TICK
+
+	total_turfs = length(turf_list)
+	log_debug("Maze at [x],[y],[z] has [total_turfs] turfs to process")
+
+	// Do the actual work
+	push_turf(turf_list[1]) // Use the first turf as the stack base
+	while(length(working_stack))
+		var/turf/T = pop_turf()
+
+		// If you dont force cast these to strings, stuff cries. A lot.
+		var/list/cardinals = list("[NORTH]", "[SOUTH]", "[EAST]", "[WEST]")
+
+		// Unvisited turfs
+		var/list/turf/unvisited_neighbours = list()
+
+		// Check all cardinal turfs
+		for(var/D in cardinals)
+			var/turf/T2 = get_step(T, text2num(D))
+			if(T2.color == MAZEGEN_TURF_UNSEARCHED)
+				unvisited_neighbours["[D]"] += T2
+
+		if(length(unvisited_neighbours))
+			push_turf(T)
+			var/D = pick(unvisited_neighbours)
+			var/turf/T3 = unvisited_neighbours["[D]"] // Pick random dir turf
+
+			// Remove the window between the two
+			for(var/obj/structure/window/reinforced/mazeglass/W in T)
```

Comment:
These cannot be made typeless, because I need only the windows, not everything in the turf like the lighting object atoms